src/libraries/System.Data.Common/src/System/Data/DataRowExtensions.cs

```diff
@@ -139,15 +140,61 @@ public static void SetField<T>(this DataRow row, DataColumn column, T? value)
             row[column] = (object?)value ?? DBNull.Value;
         }
 
-        private static T? UnboxT<T>(object value)
+        private static class UnboxT<T>
         {
-            if (value == DBNull.Value)
+            internal static readonly Func<object, T?> s_unbox = Create();
+
+            private static Func<object, T?> Create()
             {
-                if (default(T) is null)
-                    return default(T);
-                throw DataSetUtil.InvalidCast(SR.Format(SR.DataSetLinq_NonNullableCast, typeof(T)));
+                if (typeof(T).IsValueType && default(T) == null)
+                {
+                    if (!RuntimeFeature.IsDynamicCodeSupported)
+                        return NullableFieldUsingReflection;
+
+#pragma warning disable IL3050 // There is a path that is safe for AOT executed when IsDynamicCodeSupported is false.
+                    return typeof(UnboxT<T>)
+                        .GetMethod("NullableField", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic)!
+                        .MakeGenericMethod(Nullable.GetUnderlyingType(typeof(T))!)
+                        .CreateDelegate<Func<object, T>>();
+#pragma warning restore IL3050 // Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.
+                }
+                return NonNullableField;
             }
-            return (T)value;
+
+            private static T? NonNullableField(object value)
+            {
+                if (value == DBNull.Value)
+                {
+                    if (default(T) is null)
+                        return default;
+                    throw DataSetUtil.InvalidCast(SR.Format(SR.DataSetLinq_NonNullableCast, typeof(T)));
+                }
+                return (T)value;
+            }
+
+            private static T? NullableFieldUsingReflection(object value)
+            {
+                if (value == null || value == DBNull.Value)
```

Comment:
```suggestion
                if (value == DBNull.Value)
```
to match the behavior of non-reflection path when null is passed in